### ISSUE TITLE
.m4a Support

### ISSUE THREAD

**Issue by ahmetoner** [ahmetoner/whisper-asr-webservice#42]
https://github.com/openai/whisper/discussions/238#discussioncomment-3821953

**Comment by elsatch** [comment]
As a pointer for someone else having the same problems, you will get an "Internal Server Error" if you submit an m4a file.

I stumbled upon this when testing my first deployment, so I was not sure about the problem. You can diagnose your installation with a regular wav file:

```sh
curl -F "audio_file=@20221220152921_audio_note.wav" 'http://server:9000/asr?task=transcribe&language=es&output=json'
```

If this works, your installation is working! Once the support for m4a is added, it should work without problems when submitting these files too!

**Comment by jonibi** [comment]
anyone knows if original whisper code works with m4a files?

**Comment by JacoBezuidenhout** [comment]
I am experiencing the same issue... :(

ChatGPT said:

I'm sorry to hear that you are encountering an internal server error when you try to submit an m4a file to the OpenAI Whisper ASR webservice. There could be several reasons for this error, and without more information it's difficult to say for sure what the problem might be. Here are a few things you could try to troubleshoot the issue:

Check the logs for the webservice container to see if there are any error messages that might help you understand what's going wrong.

Make sure that the file you are trying to submit is a valid m4a file and that it is not corrupted or otherwise damaged.

Try submitting a different m4a file to see if the problem is with the specific file you are using.

Check the webservice configuration to ensure that it is set up to accept m4a files.

I hope these suggestions are helpful. If you continue to have trouble, please don't hesitate to ask for more help.

**Comment by JacoBezuidenhout** [comment]
Just an update, I used https://github.com/jiaaro/pydub#installation to convert the m4a files to wav and that worked well. 

def main(audio_file_url):
    url = "http://localhost:9000/asr"
    print(audio_file_url)

    if not ".m4a" in audio_file_url:
        file = requests.get(audio_file_url, stream=True).raw
    else:
        from pydub import AudioSegment
        from io import BytesIO
        # Create file-like object from raw content
        file = requests.get(audio_file_url, stream=True).raw
        filelike = BytesIO(file.read())
        file = AudioSegment.from_file(filelike, format="m4a")
        file = BytesIO(file.export(format="wav").read())


    files = {
        "audio_file": file
    }
    data = {
        "task": "transcribe",
        "language": "af",
        "output": "json",
    }
    response = requests.post(url, files=files, data=data)
    print(response.text)
    try:
        return response.json()
    except Exception as e:
        print(e)
        return response.text

**Comment by Korolen** [comment]
Just so that this shows up on Google, this is the error message I got with an .m4a:

`/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/whisper/transcribe.py:78: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")
[2023-02-23 00:48:03 -0800] [99494] [ERROR] Exception in ASGI application
Traceback (most recent call last):
  File "/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 404, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__
    return await self.app(scope, receive, send)
  File "/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/fastapi/applications.py", line 270, in __call__
    await super().__call__(scope, receive, send)
  File "/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/starlette/applications.py", line 124, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 75, in __call__
    raise exc
  File "/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 64, in __call__
    await self.app(scope, receive, sender)
  File "/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
    raise e
  File "/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
    await self.app(scope, receive, send)
  File "/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/starlette/routing.py", line 680, in __call__
    await route.handle(scope, receive, send)
  File "/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/starlette/routing.py", line 275, in handle
    await self.app(scope, receive, send)
  File "/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/starlette/routing.py", line 65, in app
    response = await func(request)
  File "/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/fastapi/routing.py", line 231, in app
    raw_response = await run_endpoint_function(
  File "/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/fastapi/routing.py", line 162, in run_endpoint_function
    return await run_in_threadpool(dependant.call, **values)
  File "/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/starlette/concurrency.py", line 41, in run_in_threadpool
    return await anyio.to_thread.run_sync(func, *args)
  File "/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/anyio/to_thread.py", line 31, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
    return await future
  File "/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 867, in run
    result = context.run(func, *args)
  File "/Users/tom/code_tmp/clone/whisper-asr-webservice/app/webservice.py", line 71, in transcribe
    result = run_asr(audio_file.file, task, language, initial_prompt)
  File "/Users/tom/code_tmp/clone/whisper-asr-webservice/app/webservice.py", line 122, in run_asr
    result = model.transcribe(audio, **options_dict)
  File "/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/whisper/transcribe.py", line 84, in transcribe
    mel = log_mel_spectrogram(audio)
  File "/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/whisper/audio.py", line 115, in log_mel_spectrogram
    stft = torch.stft(audio, N_FFT, HOP_LENGTH, window=window, return_complex=True)
  File "/Users/tom/Library/Caches/pypoetry/virtualenvs/whisper-asr-webservice-RH4kejqf-py3.10/lib/python3.10/site-packages/torch/functional.py", line 604, in stft
    input = F.pad(input.view(extended_shape), [pad, pad], pad_mode)
RuntimeError: 2D or 3D (batch mode) tensor expected for input, but got: [ torch.FloatTensor{1,1,0} ]`

**Comment by ayancey** [comment]
I think this problem is happening because when piping an MP4 container file, the moov box may be at the end of the file which is confusing ffmpeg. https://stackoverflow.com/a/62956360

**Comment by ayancey** [comment]
Indeed, if I run the input file through [qtfaststart](https://github.com/danielgtaylor/qtfaststart) first, it now seems to work.

**Comment by blundercode** [comment]
Also need this fixed. Would appreciate a merge of this. +1

**Comment by ahmetoner** [comment]
Next weekend we will have a new release with this.

**Comment by blundercode** [comment]
> Next weekend we will have a new release with this.

Great to hear thanks for the fast response. Love the repo!
If you need testers let me know!

**Comment by ahmetoner** [comment]
closed by #82
